packages/experience/src/containers/VerificationCode/index.tsx

```diff
@@ -42,15 +45,27 @@ const VerificationCode = ({ flow, identifier, className, hasPasswordButton, targ
     target
   );
 
-  useEffect(() => {
-    if (code.length === defaultLength && code.every(Boolean)) {
-      const payload =
-        identifier === SignInIdentifier.Email
-          ? { email: target, verificationCode: code.join('') }
-          : { phone: target, verificationCode: code.join('') };
-      void onSubmit(payload);
+  const [isSubmitting, setIsSubmitting] = useState(false);
+
+  const handleSubmit = useCallback(async () => {
+    if (!isCodeReady) {
+      return;
     }
-  }, [code, identifier, onSubmit, target]);
+
+    setIsSubmitting(true);
+
+    await onSubmit(
+      identifier === SignInIdentifier.Email
+        ? { email: target, verificationCode: code.join('') }
+        : { phone: target, verificationCode: code.join('') }
+    );
+
+    setIsSubmitting(false);
+  }, [code, identifier, isCodeReady, onSubmit, target]);
+
+  useEffect(() => {
+    void handleSubmit();
+  }, [handleSubmit]);
```

Comment:
this looks odd. How about passing the code into the handleSubmit callback function,  and subscribe the code change events here.

Reply:
Yeah this doesn't look like a best practice to me, either.

Reply:
updated.